smart_open/tests/test_smart_open.py

```diff
@@ -1840,6 +1841,146 @@ def test(self):
         self.assertEqual(expected, actual)
 
 
+class HandleS3CompressionTestCase(parameterizedtestcase.ParameterizedTestCase):
+    # compression | ignore_ext | behavior |
+    # ----------- | ---------- | -------- |
+    # 'gz'        | False      | Override |
+    # 'bz2'       | False      | Override |
+    @parameterizedtestcase.ParameterizedTestCase.parameterize(
```

Comment:
Will refactor.

- [ ] Refactor to simplify

Reply:
I had a hard time replicating the implementation behavior you suggested. Specifically, getting the `gzip.compress` function to produce an output that matches the results of using `gzip.GzipFile` as a stream wrapper. Here's a simple `ipython` session to illustrate what I mean:

```python
>>> from io import BytesIO
>>> import gzip
>>> 
>>> data = "не слышны в саду даже шорохи".encode("utf-8")
>>> 
>>> fh = BytesIO()
>>> file = gzip.GzipFile(fileobj=fh, mode="wb")
>>> file.write(data)
51
>>> file.close()
>>> 
>>> any([gzip.compress(data, compresslevel=level) == fh.getvalue() for level in range(1, 10)])
```

By comparison, it roundtrips cleanly, so I only slightly altered the implementation:

```python
>>> data == gzip.decompress(fh.getvalue())
True
```

Reply:
This is because you're not passing the mtime parameter (it's a timestamp). If you pass mtime, the output should be identical (have a look at how the gzip.compress function is implemented if you're unsure).

Reply:
https://github.com/python/cpython/blob/bb3e0c240bc60fe08d332ff5955d54197f79751c/Lib/gzip.py#L547-L554

Reply:
This would make the test implementation break on Python 3.7 and prior, because the extra parameter [did not exist in that version](https://github.com/python/cpython/blob/3.7/Lib/gzip.py#L529). If this is acceptable, I'll make the change.

Reply:
I take [this configuration](https://github.com/RaRe-Technologies/smart_open/blob/develop/tox.ini#L3) to mean that it would not be acceptable, so closing.

Reply:
Good catch with the python version.

You can still achieve the same effect using mock.patch, e.g.

```python
@mock.patch('time.time', return_value=123):
def gzip_compress(data):
    """For time-independent, consistent gzip compression results."""
    return gzip.compress(data)
```

Reply:
The suggestion is to compress the data *being passed* to the parameterization, which is out of scope for any mocking. The best refactoring that could be achieved would be that the `decompressor` parameter be replaced with a `compressor` parameter, so that the `gzip.compress` call is in the scope of the patched `time.time`. Either that, or the above patched method can be called outside the test method, but then must be then monkey-patched into `smart_open.open` invocation to match. Neither of those sounds like a simplification of logic, given how elaborate the patching is getting.

You're correct. It's possible. But it wouldn't be a simplification. Please consider the patch as is.